File: Problems/Algorithms/224. Basic Calculator/BasicCalculator.java
class Solution {
    public int calculate(String s) {
        Stack<Integer> stack = new Stack<>();
        
        int ans = 0, num = 0, sign = 1;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (Character.isDigit(c)) {
                num = 10 * num + (int) (c - '0');
            } else if (c == '+' || c == '-') {
                ans += sign * num;
                num = 0;
                sign = c == '+' ? 1 : -1;
            } else if (c == '(') {
                stack.push(ans);
                stack.push(sign);
                // num = 0;
                ans = 0;
                sign = 1;
            } else if (c == ')') {
                ans += sign * num;
                num = 0;
                ans *= stack.pop();
                ans += stack.pop();
            }
        }
        
        if (num != 0) {
            ans += sign * num;
        }
        
        return ans;
    }
}

File: Problems/Algorithms/224. Basic Calculator/basic_calculator.py
#!/usr/bin/env python3
# coding: utf-8


class Solution:
    def calculate(self, s: str) -> int:
        stack = []

        ans, num, sign = 0, 0, 1
        for c in s:
            if c.isdigit():
                num = num * 10 + int(c)
            elif c in '+-':
                ans += sign * num
                num = 0
                sign = 1 if c == '+' else -1
            elif c == '(':
                stack.append(ans)
                stack.append(sign)
                ans, sign = 0, 1
            elif c == ')':
                ans += sign * num
                ans *= stack.pop()
                ans += stack.pop()
                num = 0
        
        return ans + sign * num
